src/librustc_typeck/collect.rs

```diff
@@ -1682,28 +1682,65 @@ fn find_existential_constraints(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
     }
 }
 
+pub fn get_infer_ret_ty(output: &'_ hir::FunctionRetTy) -> Option<&hir::Ty> {
+    if let hir::FunctionRetTy::Return(ref ty) = output {
+        if let hir::TyKind::Infer = ty.node {
+            return Some(&**ty)
+        }
+    }
+    None
+}
+
 fn fn_sig(tcx: TyCtxt<'_>, def_id: DefId) -> ty::PolyFnSig<'_> {
     use rustc::hir::*;
     use rustc::hir::Node::*;
 
     let hir_id = tcx.hir().as_local_hir_id(def_id).unwrap();
 
-    let icx = ItemCtxt::new(tcx, def_id);
-
     match tcx.hir().get(hir_id) {
         TraitItem(hir::TraitItem {
-            node: TraitItemKind::Method(sig, _),
+            node: TraitItemKind::Method(MethodSig { header, decl }, TraitMethod::Provided(body_id)),
             ..
         })
         | ImplItem(hir::ImplItem {
-            node: ImplItemKind::Method(sig, _),
+            node: ImplItemKind::Method(MethodSig { header, decl }, body_id),
+            ..
+        })
+        | Item(hir::Item {
+            node: ItemKind::Fn(decl, header, _, body_id),
             ..
-        }) => AstConv::ty_of_fn(&icx, sig.header.unsafety, sig.header.abi, &sig.decl),
+        }) => {
+            if let Some(ty) = get_infer_ret_ty(&decl.output) {
+                let tables = tcx.typeck_tables_of(def_id);
+                let mut diag = tcx.sess.struct_span_err_with_code(
+                    ty.span,
+                    "the type placeholder `_` is not allowed within types on item signatures",
+                    DiagnosticId::Error("E0121".into()),
+                );
+                diag.span_label(ty.span, "not allowed in type signatures");
+                if let Some(ret_ty) = tables.node_type_opt(body_id.hir_id) {
+                    diag.span_suggestion(
+                        ty.span,
+                        "replace `_` with the correct return type",
+                        format!("{}", ret_ty),
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+                diag.emit();
```

Comment:
cc @estebank is this use of diagnostics correct?

Reply:
It'd be nice if we could have `ty_infer` and this branch call the same code that would generate `E0121` to have a single point where the error is created. Just making a function with the following should be enough:

```
            fn bad_placeholder_type(&self, tcx: &TyCtxt<'_>) -> DiagnosticBuilder<'_> {
                let mut diag = tcx.sess.struct_span_err_with_code(
                    ty.span,
                    "the type placeholder `_` is not allowed within types on item signatures",
                    DiagnosticId::Error("E0121".into()),
                );
                diag.span_label(ty.span, "not allowed in type signatures");
                diag
            }
```

Then this fn can be called from both places.

Finally, I believe that this is missing a call to `resolve_type_vars_with_obligations`, in case the fn body's resolved type is something like `Option<T>`. I would like to see a test like:

```
fn foo() -> _ {
    let x = Some(42);
    x
}

fn main() {
    let _: Option<usize> = foo();
    let _: f64 = foo();
    let _: Option<i32> = foo();
}
```

Beyond that, I love this PR ❤️